drivers/adc/adc_npcx.c

```diff
@@ -316,6 +684,8 @@ static const struct adc_npcx_config adc_npcx_cfg_0 = {
 	.alts_list = adc_alts,
 };
 
+static struct adc_npcx_threshold_data threshold_data_0;
```

Comment:
If we go with this choice we would need to also guard statement `data->threshold_data = &threshold_data_0;`, otherwise we have build errors whenever CONFIG_ADC_CMP_NPCX is not defined. Same applies with function `adc_npcx_set_repetitive`.

Moving threshold implementation to other file would require to pass members of `struct adc_npcx_data` like `ctx` to obtain semaphore when registers are access. Also we would need to guard code intended to preserve ADC functionality and coexist with threshold feature.

Reply:
My concern is that the threshold support as implemented increases the size of the NPCX ADC driver, even when `CONFIG_ADC_CMP_NPCX=n`.  Can you check this?  I'm not certain how much the linker will strip out within a module when LTO (link time optimization) is not enabled.

I see that the assignment of `data->threshold-data` is already guarded with the Kconfig option.  This would just need to be switched to an #ifdef style guard.

Reply:
Let me do some objdump and confirm if functions are being included in binaries when `CONFIG_ADC_CMP_NPCX=n`.

I am also considering moving all this code into an adc_npcx_threshold.c, though I am not sure that it would be accepted to have single peripheral set of registers accessed in more than one file.

Reply:
I compared a couple of builds, both targeted for npcx and ADC enabled. Despite not having this code guarded with `#ifdef`, linker is doing a good job not including functions and `threshold_data_0` variable.
- `CONFIG_ADC_CMP_NPCX=y`
```
00000000 l    df *ABS*  00000000 adc_npcx.c
100aa372 l     F text   00000078 adc_context_start_sampling
100aa3ea l     F text   00000004 k_sem_give
100aa3ee l     F text   00000006 z_log_msg2_static_create.constprop.0
100a3038 l     F text   000000cc adc_npcx_channel_setup
100aa3f4 l     F text   00000174 adc_npcx_isr
100aa568 l     F text   0000002e adc_context_on_timer_expired
100a3104 l     F text   00000114 adc_npcx_init
100aa596 l     F text   0000000e adc_context_lock.constprop.0
100a3218 l     F text   0000016c adc_npcx_read
**200c7474 l     O bss    00000034 threshold_data_0**
200c0990 l     O datas  000000c0 adc_npcx_data_0
100b85af l     O rodata 00000018 adc_alts
100b004c l     O rodata 0000000c adc_npcx_cfg_0
100b0058 l     O rodata 0000000c adc_npcx_driver_api
```
- `CONFIG_ADC_CMP_NPCX=n`
```
00000000 l    df *ABS*  00000000 adc_npcx.c
100a7d38 l     F text   00000078 adc_context_start_sampling
100a7db0 l     F text   00000042 adc_npcx_channel_setup
100a7df2 l     F text   00000034 adc_context_on_timer_expired
100a7e26 l     F text   0000010c adc_npcx_read
100a7f32 l     F text   00000104 adc_npcx_isr
100a3ae8 l     F text   000000c4 adc_npcx_init
200c0cf8 l     O datas  000000c0 adc_npcx_data_0
100b4e4a l     O rodata 00000018 adc_alts
100ae638 l     O rodata 0000000c adc_npcx_cfg_0
100ae644 l     O rodata 0000000c adc_npcx_driver_api
```

I am trying to disable optimizations to make sure we have same results.